### PR TITLE
[RFR] Add custom caching decorator and implement on various properties

### PR DESCRIPTION
**Pull request by cwisecarver**
## Purpose

Make less guid queries

## Changes

Add custom cached_property decorator that supports setters. Implement on GuidMixin._id and Node.parent_node and Node.root.

## Side effects

Could cause setting guids on existing objects to not work good.

### CODE REVIEW

**Comment by cwisecarver** [comment]
Passed on my local TCI.

**Comment by cwisecarver** [comment]
https://travis-ci.org/cwisecarver/osf.io/builds/207507731

File: osf/utils/caching.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+The cache is stored on the model as a protected attribute. Expensive
+property lookups, such as database access, can therefore be sped up
+when accessed multiple times in the same request.
+The property can also be safely set and deleted without interference.
+"""
+from __future__ import unicode_literals
+
+from functools import wraps
+
+# from https://github.com/etianen/django-optimizations/blob/master/src/optimizations/propertycache.py
```
Comment (sloria) [inline osf/utils/caching.py]: This code is BSD-3 licensed. Might need to put the license text in our `NOTICE` file: https://github.com/etianen/django-optimizations/blob/master/LICENSE

Comment (cwisecarver) [inline osf/utils/caching.py]: Will do.

**Pull request merged**